### PR TITLE
Update ActiveMQ to 5.16.4

### PR DESCRIPTION
**Pull request by brudo**
This updates ActiveMQ to the recently released version 5.16.4, to get the latest and most stable version of ActiveMQ Client that runs on both JRE8 and later JREs. It replaces of log4j 1.2.x with reload4j, although not sure if this affects the client.

In addition, slf4j is updated to the latest version, which uses reload4j in lieu of log4j 1.2.x, and an exclusion is added to the GeoEvent SDK dependency to ensure that the original log4j 1.2.x does not get pulled at build time. Finally, an explicit dependency for reload4j was also added, just to be sure.

After this change, I confirmed that log4j no longer appears in the compile-time dependencies as shown in NetBeans; and the classes for the explicitly selected versions of slf4j and reload4j are present in the resulting JAR, ensuring that ActiveMQ for GeoEvent will use these classes in preference to any that may be provided by GeoEvent Server itself.

It is unknown whether the use of ActiveMQ for GeoEvent could ever lead to any problems with log4j 1.2.x vulnerabilities. However, many organizations have a goal to eliminate any chance of exposure to known log4j 1.2.x and 2.x vulnerabilities.

Incidentally, ActiveMQ 5.17 has just been released as well, with many bug fixes and improvements. However, we can't use that in ActiveMQ for GeoEvent, because GeoEvent Server 10.8.x and earlier run under JRE8, which is no longer supported for ActiveMQ 5.17. (I know this from the mailing lists; it doesn't seem to be prominently mentioned in the release announcement.)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -27,24 +28,40 @@
...
+        <exclusion>
+          <groupId>log4j</groupId>
+          <artifactId>log4j</artifactId>
+        </exclusion>
+      </exclusions> 
+    </dependency>
+    <dependency>
+        <groupId>ch.qos.reload4j</groupId>
+        <artifactId>reload4j</artifactId>
+        <version>1.2.19</version>
```
Comment (hanoch) [inline pom.xml]: Can you try replacing that with a subset of (and scope `provided` when you can):
```
org.slf4j:slf4j-api:1.7.32 --> you can try v1.7.35 as well
org.slf4j:slf4j-log4j12:1.7.32 --> only if needed.
org.apache.logging.log4j:log4j-slf4j-impl:2.17.1 -> only if needed
org.apache.logging.log4j:log4j-core:2.17.1
org.apache.logging.log4j:log4j-api:2.17.1
org.apache.logging.log4j:log4j-1.2-api:2.17.1 --> this is the log4j to log4j2 API bridge
```

**Review by hanoch — CHANGES_REQUESTED** [review]
Please install ArcGIS GeoEvent Server 10.8.1 Patch 4 (https://support.esri.com/en/download/8000), which includes the log4shell security fixes, and try to do the following:

File: pom.xml (hunk 1 above)
Comment (brudo) [reply pom.xml]: Keep in mind that this component supports GeoEvent Server 10.4.1 and above, and is built against the SDK as published on Maven Central, independent of what patches are installed on the development machine.

However, if you are concerned it would cause a problem to have those files embedded, I'm OK with changing the dependency to provided. Then, at run time it relies on the underlying GeoEvent Server to be patched and/or stripped of the problematic classes. I just wasn't sure what would happen if it was actually the original log4j:log4j that was provided. 
@hanoch do you think that would be better?

For slf4j, I understood that it was at v1.7.35 where they initially switched to reload4j; however, as v1.7.36 was released a month later, and uses the latest reload4j 1.2.19, I thought probably best to use that one.

Comment (brudo) [reply pom.xml]: Actually, perhaps best to mark reload4j as both provided and also optional, and also specify that slf4j is provided, even though a version is specified. Then neither module is included in the JAR, and it is up to GeoEvent Server to provide these classes at runtime, as before. (I'm not sure if the 'optional' is strictly needed when scope is 'provided' anyway, but thought it might be safer / clearer, since the classes might be supplied by log4j rather than reload4j in an older version.)

File: pom.xml [hunk 2]
```diff
@@ -27,24 +28,43 @@
...
         <artifactId>commons-logging</artifactId>
-        <version>1.1.1</version>
+        <version>1.2</version>
+      </dependency>
+      <dependency>
+          <groupId>org.slf4j</groupId>
+          <artifactId>slf4j-api</artifactId>
+          <version>1.7.36</version>
+          <scope>provided</scope>
       </dependency>
```
Comment (hanoch) [inline pom.xml]: This might not be needed, since it is already a transitive dep of ActiveMQ.
Please remove this new dep and test again without it.

File: pom.xml [hunk 3]
```diff
@@ -27,24 +28,43 @@
...
+        <exclusion>
+          <groupId>log4j</groupId>
+          <artifactId>log4j</artifactId>
+        </exclusion>
+      </exclusions> 
+    </dependency>
+    <dependency>
+        <groupId>ch.qos.reload4j</groupId>
+        <artifactId>reload4j</artifactId>
+        <version>1.2.19</version>
```
Comment (hanoch) [inline pom.xml]: Please revert this change and test with the following instead:
```
    <dependency>
      <groupId>com.esri.geoevent.sdk</groupId>
      <artifactId>geoevent-sdk</artifactId>
      <version>${project.version}</version>
      <scope>provided</scope>
      <exclusions>
        <exclusion>
          <groupId>org.ops4j.pax.logging</groupId>
          <artifactId>*</artifactId>
        </exclusion>
        <exclusion>
          <groupId>org.slf4j</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
```

**Review by hanoch — CHANGES_REQUESTED** [review]
Please retest with the following suggestions.
(Which version of GeoEvent Server are you using for testing?)
We need to refrain from adding more direct dependencies since it will force us to start maintaining their versions. It is better to let them come in naturally as transitive dependencies, as much as possible:

**Comment by brudo** [comment]
I tried the change you requested.
 * The pax.logging exclusion doesn't seem to make any difference at compile time or run time.
 * The slf4j exclusion would also have to be included on activemq, as that is how it's ending up in the JAR
 * I don't know why wildcard exclusions were needed, and thought they might be too heavy handed.

I also found that, depending on the version of GeoEvent SDK, the version of commons-logging may be 1.1.1 (for 10.4.1) or 1.2 (for 10.8.1). I'm not sure why that was specified in dependencyManagement at all, but removing it doesn't seem to change anything.

However, I think the following is clearer, and easier to maintain: if I just have this in dependencyManagement, it seems to be enough, ensuring that no version of slf4j is embedded in the JAR and we rely entirely on GeoEvent Server to provide a good version of it:

```
<dependencyManagement>
    <dependencies>
      <dependency>
          <groupId>org.slf4j</groupId>
          <artifactId>slf4j-api</artifactId>
          <scope>provided</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>
```

Then there are no exclusions needed in geoevent-sdk or activemq-client, and no need to reference reload4j at all.

Note that it was activemq-client that was causing the slf4j dependency to appear in the JAR originally, not geoevent-sdk, since the geoevent-sdk itself was marked as provided.

Of course, we do rely on GeoEvent Server to be patched / vulnerabilities mitigated.

**Comment by hanoch** [comment]
My suggestion was to try to remove any dep on older (than 2.17.1) log4j versions coming from the SDK and let the ActiveMQ lib bring its own.
Your suggestion works too, from the point of view of as lease maintenance as possible and as many different versions support as possible.

File: pom.xml [hunk 4]
```diff
@@ -15,7 +15,8 @@
     <contact.address>geoevent@esri.com</contact.address>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <maven.bundle.plugin.version>4.2.1</maven.bundle.plugin.version>
-    <activemq.version>5.16.0</activemq.version>
+    <activemq.version>5.16.4</activemq.version>
+    <!-- can't go to 5.17 if GeoEvent is running under JRE8 - which 10.8.x and earlier do -->
```
Comment (hanoch) [inline pom.xml]: can you rename `GeoEvent` to `ArcGIS GeoEvent Server` in the comment here?
`GeoEvent` is the actual event (message), not the official product name.

File: pom.xml [hunk 5]
```diff
@@ -25,12 +26,13 @@
   <dependencyManagement>
     <dependencies>
       <dependency>
-        <groupId>commons-logging</groupId>
-        <artifactId>commons-logging</artifactId>
-        <version>1.1.1</version>
+          <groupId>org.slf4j</groupId>
+          <artifactId>slf4j-api</artifactId>
+          <scope>provided</scope>
```
Comment (hanoch) [inline pom.xml]: please make sure to match the tabs/spaces file indentation.

**Review by hanoch — APPROVED** [review]

**Comment by brudo** [comment]
@hanoch OK, let's leave it as I have it now; I think it's clear what it will do, meets your requirements, and is nicely concise.

Note that I do not have write access, so someone else will have to merge it.

**Comment by hanoch** [comment]
@brudo please make the above two changes.
Did you test this PR?  If so - against which version of GeoEvent Server?

**Review by hanoch — CHANGES_REQUESTED** [review]
@brudo I can merge this PR.
Please make the above two changes (comment adjustment + tabs/spaces).
Did you test these changes?  If so - against which version of ArcGIS GeoEvent Server?

File: pom.xml (hunk 4 above)
Comment (brudo) [reply pom.xml]: OK

File: pom.xml (hunk 5 above)
Comment (brudo) [reply pom.xml]: Fixed

**Review by hanoch — APPROVED** [review]

**Comment by brudo** [comment]
> Did you test these changes? If so - against which version of ArcGIS GeoEvent Server?

Yes, it was tested - primarily in 10.7.1 but also to confirm it would load in 10.8.1. I expect it should work with 10.4.1 - 10.8.1.

However, I believe a new revision will likely be required for 10.9.x and above, as it would be better to target JDK 11 starting from 10.9.x, and there seems to be a breaking change in the GeoEvent SDK (`error: cannot access LabeledValue` when I try to compile against 10.9.0 or 10.9.1) which needs to be remedied (one way would be to switch to an XML transport definition). I will probably take that on, and provide a separate pull request once ready.
